### PR TITLE
ARROW-6962: [C++] [CI] Stop compiling with -Weverything

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
I'll be inclined to merge this as soon as CI passes (or the errors seem unrelated). @wesm @kou

**Comment by pitrou** [comment]
Even with this PR, `-Weverything` still crops up in the compile options: https://travis-ci.org/pitrou/arrow/jobs/601324968#L1650 :-/

**Comment by pitrou** [comment]
Well, of course, it's ORC adding these options:
https://github.com/apache/orc/blob/master/CMakeLists.txt

**Comment by pitrou** [comment]
Perhaps we should add `-Wextra` as well for "checkin" builds. But will check first whether builds without it pass.

**Review by bkietz — APPROVED** [review]

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 1]
```diff
@@ -100,9 +100,8 @@ endif()
 # BUILD_WARNING_LEVEL add warning/error compiler flags. The possible values are
 # - PRODUCTION: Build with `-Wall` but do not add `-Werror`, so warnings do not
 #   halt the build.
-# - CHECKIN: Imply `-Weverything` with certain pedantic warnings
-#   disabled. `-Werror` is added in debug builds so that any warnings fail the
-#   build
+# - CHECKIN: Build with `-Wall` (and `-Werror` in debug mode so that any
```
Comment (kou) [inline cpp/cmake_modules/SetupCxxFlags.cmake]: Could you add `-Wextra` here too?

**Review by kou — APPROVED** [review]

File: cpp/cmake_modules/SetupCxxFlags.cmake (hunk 1 above)
Comment (pitrou) [reply cpp/cmake_modules/SetupCxxFlags.cmake]: Will do.

**Comment by pitrou** [comment]
Travis-CI: https://travis-ci.org/pitrou/arrow/builds/601514202

**Comment by wesm** [comment]
If we aren't using `-Weverything`, then we probably don't need to have a long list of disabled warnings. 

Does `-Wall` on clang catch missing virtual destructors? We may need to opt in to some of the warnings that we care about.

**Pull request closed without merging**